torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,159 @@
+    strides = _strides.value();
+  } else {
+    strides = at::detail::defaultStrides(sizes);
+  }
+  return std::make_pair(sizes, strides);
+}
+
+size_t computeStorageSize(const c10::TensorTypePtr& ttp) {
+  std::vector<int64_t> sizes, strides;
+  std::tie(sizes, strides) = getSizesStrides(ttp);
+  return at::detail::computeStorageNbytes(
+      sizes, strides, elementSize(ttp->scalarType().value()));
+}
+
+std::pair<size_t, std::map<Value*, c10::TensorTypePtr>>
+computeAlignedStorageSize(std::vector<Node*>& nodes) {
+  // TODO: liveness analysis
+  std::map<Value*, c10::TensorTypePtr> managed_tensors;
+  size_t managed_bytes = 0;
+  for (const auto& node : nodes) {
+    for (const auto& output : node->outputs()) {
+      auto ttp = checkIsSupported(*output);
+      managed_bytes += computeStorageSize(ttp);
+      managed_tensors.insert({output, ttp});
+    }
+  }
+
+  auto aligned_size = MemoryPlanner::compute_aligned_tensor_size(managed_bytes);
+  return std::make_pair(aligned_size, managed_tensors);
+}
+
+DeviceType findDeviceType(std::shared_ptr<Graph>& graph) {
+  for (const auto& inp : graph->inputs()) {
+    auto typ = inp->type()->cast<TensorType>();
+    if (typ && typ->device().has_value()) {
+      return typ->device().value().type();
+    }
+  }
+
+  for (const auto& node : graph->nodes()) {
+    for (const auto& inp : node->inputs()) {
+      auto typ = inp->type()->cast<TensorType>();
+      if (typ && typ->device().has_value()) {
+        return typ->device().value().type();
+      }
+    }
+  }
+  TORCH_CHECK(false, "couldn't find device type")
+}
+
+Node* insertAllocStorageNode(std::shared_ptr<Graph>& graph, size_t total_size) {
+  auto deviceType = findDeviceType(graph);
+  auto* storage = graph->create(prim::AllocateStorage, 1);
+  // TODO: pass device type here
+  storage->i_(attr::total_size, total_size);
+  storage->i_(attr::device, static_cast<int8_t>(deviceType));
+  storage->insertBefore(graph->nodes().front());
+  return storage;
+}
+
+void insertAllocTensorNodes(
+    std::shared_ptr<Graph>& graph,
+    std::vector<Node*> out_nodes,
+    Node* storage,
+    std::map<Value*, c10::TensorTypePtr>& managed_tensors) {
+  size_t offset = 0;
+  size_t total_size = storage->i(attr::total_size);
+  for (auto* node : out_nodes) {
+    auto* alloc = graph->create(prim::AllocateTensor, 1);
+    node->addInput(alloc->output());
+    GRAPH_DEBUG("inserting allocation op for ", node->getOperator().schema());
+    alloc->insertBefore(node);
+    alloc->addInput(storage->output());
+
+    TORCH_CHECK(
+        managed_tensors.count(node->output()) > 0,
+        "output value not managed by memory planner.");
+    auto ttp = managed_tensors.at(node->output());
+    auto size = computeStorageSize(ttp);
+    std::vector<int64_t> sizes, strides;
+    std::tie(sizes, strides) = getSizesStrides(ttp);
+    TORCH_CHECK(offset + size <= total_size, "not enough memory");
+    alloc->i_(attr::size, size);
+    alloc->i_(attr::offset, offset);
```

Comment:
do we have alignment requirements for individual tensors? 
say, the first tensor is `torch.rand(1, dtype=torch.int8)` we probably don't want the second tensor start on a byte boundary?

Reply:
@ngimel I think you might know about this...

Reply:
We don't have requirements other than alignment should be multiple of `sizeof(type)`, but perf will be better if alignment is better (e.g. vectorization on gpu will kick in more or less when alignment is `4*sizeof(type)`, cache lines are 128 bytes so with 128 bytes alignment you are getting better cache utilization, tensor cores can be used with, I don't remember, 256 or 512 bytes alignment?). CUDA caching allocator I think lines all new allocations to 512 bytes.